Mini Cart: stop using Modal component

Aljullu opened the pull request:
This PR removes the usage of the `Modal` component from `@wordpress/components`. Instead, it updates our `Drawer` component to incorporate some functionality from `Modal`. I didn't copy & paste everything from the `Modal` component, instead, I just tried to import the features that were needed for our use case.

Fixes #8606.

Part of #8452.

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Testing

#### User Facing Testing

This PR shouldn't introduce any visual changes for users, so testing it means making sure there are no regressions.

1. Add the Mini Cart block to the header of your site via (Appearance > Editor). 
2. In the frontend and with the Cart empty, open the Mini Cart drawer, verify you can open and close the drawer without problems.
3. Add some products to your cart and open the Mini Cart drawer again. Verify you can open and close it, you can change the products quantity, etc.
4. Go to Appearance > Editor > Template Parts > Mini Cart and change some of the styles of the inner blocks. For example, add a custom background, border and width to the Mini Cart Contents block.
5. Repeat step 3.

Now, let's test that things keep working if there are two Mini Cart blocks in the same page. We don't officially support it, but at the same time we don't want the experience to be broken.

6. create a post or page and add the Mini Cart block.
7. Open that post/page in the frontend. You should now have two Mini Cart buttons in the screen, the one from the site header and the one from the post/page.
8. Verify both buttons work as expected.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Mini Cart block no longer uses the Modal component from `@wordpress/components`.

Aljullu committed:
Remove unused styles

Aljullu committed:
Replace usage of Modal component with custom Drawer

Aljullu committed:
Update MiniCart.php class structure

Aljullu commented on src/BlockTypes/MiniCartContents.php:
```diff
@@ -74,15 +74,6 @@ protected function enqueue_assets( array $attributes ) {
 		$bg_color   = StyleAttributesUtils::get_background_color_class_and_style( $attributes );
 
 		$styles = array(
-			array(
-				'selector'   => '.wc-block-mini-cart__drawer .components-modal__header',
-				'properties' => array(
-					array(
-						'property' => 'color',
-						'value'    => $text_color ? $text_color['value'] : false,
-					),
-				),
-			),
```
I think these styles weren't used anywhere, as we weren't setting a modal header.

Aljullu commented on assets/js/base/components/drawer/style.scss:
```diff
@@ -109,44 +109,48 @@ $drawer-animation-duration: 0.3s;
-			@include visually-hidden();
-		}
-	}
+.admin-bar .wc-block-components-drawer__content {
```
This has been moved from here:

https://github.com/woocommerce/woocommerce-blocks/blob/f1714a70bb68d7bae57e7e8a460684c6c5f031ed/assets/js/blocks/mini-cart/style.scss#L228-L238

Aljullu committed:
Update tests

thealexandrelara reviewed (changes requested):
✅ I have no complaints about the code
✅ I can open and close the Mini Cart drawer when the cart is empty
✅ I can open and close the Mini Cart drawer when the cart has products
✅ I can customize the styles for the Mini Cart (adding custom backgrounds, border, and/or changing the width)
✅ I can add multiple Mini Cart drawers and they work together

⚠️ When I open the Mini Cart drawer, the Close button appears always in a "focus" state which I think is valid from an accessibility point of view, but just want to confirm if the style is correct and we should, by default, show the "blue border" around the Close button.

![image](https://github.com/woocommerce/woocommerce-blocks/assets/20469356/84a357fd-1c87-42b8-8e0a-222f4b59eefd)

⛔ When you decrease an specific item quantity from 2 to 1 (it looks like it only happens between these quantities) by clicking on the decrease button for this item, the drawer does not close when you click outside its area. The only way I managed to close it (besides using the Close button) was to give a click inside the drawer area followed by a click outside it as in this video:

https://github.com/woocommerce/woocommerce-blocks/assets/20469356/922f81c1-6856-41ac-8e0c-0f2d50c858cf

Aljullu committed:
Prevent focus styles to appear unnecessarily when opening the Mini Cart drawer

Aljullu committed:
Work-around issue with disabled buttons causing lose of focus inside the Mini-Cart drawer

Aljullu commented:
> :warning:  When I open the Mini Cart drawer, the Close button appears always in a "focus" state which I think is valid from an accessibility point of view, but just want to confirm if the style is correct and we should, by default, show the "blue border" around the Close button.

Good point. In bf1f53cb077c3f356dfcbebbcdbd147703ba28bd I added some styles to reset the default `:focus` styles if the close button doesn't match the `:focus-visible` pseudo-class. In my testing it seems to work well: opening the drawer with the keyboard shows the focus styles, but opening it with the mouse doesn't until the close button is focused again.

Could you test it and see if it looks good to you? :pray: 

> :no_entry:  When you decrease an specific item quantity from 2 to 1 (it looks like it only happens between these quantities) by clicking on the decrease button for this item, the drawer does not close when you click outside its area. The only way I managed to close it (besides using the Close button) was to give a click inside the drawer area followed by a click outside it as in this video:

Good catch! That was a super confusing one, and I still don't understand what's going on. When the `-` button gets disabled, the focus returns to the `<body>`, which causes the drawer to close (in Firefox) or the drawer not possible to close clicking outside (Chrome). I don't know why the focus is lost, though, as I tried different approaches (setting a `key`, debouncing the `disabled` state...) and none of them worked. What's even more confusing is that if I remove the text input field, the issue gets fixed magically. :confused: 

In the end, I decided for not setting the `disabled` attribute and instead simulate it via `tabIndex`, `aria-hidden` and CSS. You can see it here: 988b7977f4e620caf0b977bf9683e10349635a03. I'm not very happy with it because it's a work-around instead of a real fix, but during my testing it performed well and I couldn't find any other solutions.

thealexandrelara reviewed (approved):
Great job, thank you for addressing the issues, now I can confirm that it is working well. I imagine how difficult it was to debug and find a solution for this, I'm glad that you find a way to solve it 🙌 

I'm approving this PR since it is working as expected, but maybe a suggestion that I don't know if you tried is to keep a reference to the modal and when clicking on the QuantitySelector, you set the focus back to the modal, something like this:

```

// on Mini Cart
const modalRef = React.useRef();

<Drawer ref={modalRef} />

// on QuantitySelector if we find a way to send the modal reference or maybe to call a callback inside the onClick for the decrease button

<button 
  ...
  onClick={ () => {
    ...
    modalRef.current.focus(); // or a callback that calls the focus() method
  }}
/>
```

Aljullu committed:
Revert "Work-around issue with disabled buttons causing lose of focus inside the Mini-Cart drawer"

This reverts commit 4360f77384ad1f1d90a3ba8a0385ad79da2449f2.

Aljullu committed:
QuantitySelector: focus text input field after decrease or increase button become disabled

Aljullu commented:
> I'm approving this PR since it is working as expected, but maybe a suggestion that I don't know if you tried is to keep a reference to the modal and when clicking on the QuantitySelector, you set the focus back to the modal, something like this:

Oh, that's a good idea! I reverted my previous work-around in 2274917eb945cd76713e9c8462fb37b9f2260265 and implemented your suggestion in 4d75aa377255b2d91f1140484909835e72b24d92. I changed it a bit so instead of moving the focus to the drawer, we move it to the text input field. This keeps the code cleaner and from an accessibility point of view I think it's better because it moves the focus closer to the previous position.

Thanks so much for the suggestion, @thealexandrelara! Would you mind taking another look?

thealexandrelara commented:
Thank you for implementing the suggestion 🙌 . I was testing it again, and it looks like the issue is still happening on Chrome (I'm using version 113.0.5672.92)

https://github.com/woocommerce/woocommerce-blocks/assets/20469356/462449af-d260-41c1-b1b2-01e7c85a605e

I tested on Firefox and Safari and the drawer closes when clicking outside it 🤔

Aljullu committed:
Move focus to the input field also when the body has focus, that fixes the issue in Chrome

Aljullu committed:
Add explanatory comment

Aljullu commented:
Thanks for the review, @thealexandrelara, and good catch with that issue. Honestly, I don't know why that's happening, for some reason, Chrome is moving the focus to `<body>` even before `useLayoutEffect()` runs. :confused: 

I solved it by adding a check to see whether the focus is in the `<body>` element: 137770b7e61c6665f1270dcf5c29f879a83d70ac. I also pushed a commit which adds a comment explaining that part of the code for future reference: 7f59f056cc498cad94ef7400cab029d344c0a19e.

thealexandrelara reviewed (approved):
Awesome! Thank you for your hard work on this one, just tested and it is working as expected 🙌. LGTM! 🚀
